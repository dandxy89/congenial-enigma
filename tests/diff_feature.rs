#![cfg(feature = "diff")]
#![cfg(feature = "serde")]

use std::path::PathBuf;

use lp_parser_rs::{
    model::{coefficient::Coefficient, constraint::Constraint, lp_error::LPParserError, lp_problem::LPProblem, sense::Cmp},
    parse::{parse_file, parse_lp_file},
};

#[test]
fn test_coefficient() {
    use diff::Diff;

    let a = Coefficient { var_name: "a".to_string(), coefficient: 1.0 };
    let b = Coefficient { var_name: "b".to_string(), coefficient: 1.0 };
    insta::assert_yaml_snapshot!(a.diff(&a));
    insta::assert_yaml_snapshot!(a.diff(&b));
}

#[test]
fn test_constraint() {
    use diff::Diff;

    let a = Constraint::Standard {
        name: "a".to_string(),
        coefficients: vec![Coefficient { var_name: "ca".to_string(), coefficient: 1.0 }],
        sense: Cmp::GreaterThan,
        rhs: 1.0,
    };
    let b = Constraint::Standard {
        name: "b".to_string(),
        coefficients: vec![Coefficient { var_name: "ca".to_string(), coefficient: 1.0 }],
        sense: Cmp::GreaterThan,
        rhs: 1.0,
    };
    insta::assert_yaml_snapshot!(a.diff(&a));
    insta::assert_yaml_snapshot!(a.diff(&b));
}

#[test]
fn test_file_comparison() {
    use diff::Diff;

    let a = read_file_from_resources("test.lp").unwrap();
    let a_copy = read_file_from_resources("test_copy.lp").unwrap();
    let b = read_file_from_resources("test2.lp").unwrap();

    insta::assert_yaml_snapshot!(a.diff(&a), {
        ".variables.altered" => insta::sorted_redaction(),
        ".variables.removed" => insta::sorted_redaction(),
        ".objectives.Altered" => insta::sorted_redaction(),
        ".objectives.Altered" => insta::sorted_redaction(),
        ".constraints.altered" => insta::sorted_redaction(),
        ".constraints.removed" => insta::sorted_redaction(),
    });

    insta::assert_yaml_snapshot!(a.diff(&a_copy), {
        ".variables.altered" => insta::sorted_redaction(),
        ".variables.removed" => insta::sorted_redaction(),
        ".objectives.Altered" => insta::sorted_redaction(),
        ".objectives.Altered" => insta::sorted_redaction(),
        ".constraints.altered" => insta::sorted_redaction(),
        ".constraints.removed" => insta::sorted_redaction(),
    });

    insta::assert_yaml_snapshot!(a.diff(&b), {
        ".variables.altered" => insta::sorted_redaction(),
        ".variables.removed" => insta::sorted_redaction(),
        ".objectives.Altered" => insta::sorted_redaction(),
        ".objectives.Altered" => insta::sorted_redaction(),
        ".constraints.altered" => insta::sorted_redaction(),
        ".constraints.removed" => insta::sorted_redaction(),
    });
}

fn read_file_from_resources(file_name: &str) -> Result<LPProblem, LPParserError> {
    let mut file_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    file_path.push(format!("resources/{file_name}"));
    let contents = parse_file(&file_path)?;
    parse_lp_file(&contents)
}
