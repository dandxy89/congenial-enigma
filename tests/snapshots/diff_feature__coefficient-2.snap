---
source: tests/diff_feature.rs
expression: a.diff(&b)
---
var_name: b
coefficient: 0
